gh-130090: Support PGO for clang-cl

chris-eibl opened the pull request:
Per encouragement from @Fidget-Spinner in https://github.com/faster-cpython/ideas/issues/690#issuecomment-2646342506 opened as a draft PR:

Support PGO for clang-cl on Windows using a similar approach as done in the Linux makefiles for clang.
I separate the clang-cl profiles into e.g. `obj\314amd64_PGInstrument\__clang_profiles`, so that they don't clash with different build configurations.

Since pythoncore is built first and not in parallel, this was the easiest spot for me.
Fore sure room for improvements, I am definitely far from being an msbuild / vcxproj expert :)

BTW, clang-cl PGO takes far less time to build compared to MSVC, because
- running the instrumented pgo tests is much faster
- the PGO step is faster

Some thoughts:

- Linux uses `-flto-thin` to improve build time over execution time. Tried it, doesn't cost much in run time, but speeds up the build a lot.
- There are still many warnings. I've fixed many of them, if interested, we can do them in follow-up PRs. IMHO, some of them are not benign, e.g.
```
..\PC\launcher2.c(1081,34): warning : result of comparison of constant 191 with expression of type 'char' is always false [-Wtautological-constant-out-of-range-compare] [e:\cpython_clang\PCbuild\pywlauncher.vcxproj]
..\PC\launcher2.c(1081,18): warning : result of comparison of constant 187 with expression of type 'char' is always false [-Wtautological-constant-out-of-range-compare] [e:\cpython_clang\PCbuild\pywlauncher.vcxproj]
``` 
- let clang-cl benefit from `_Py_HOT_FUNCTION`, etc. Most probably in follow up PRs (very simple and small)

<!-- gh-issue-number: gh-130090 -->
* Issue: gh-130090
<!-- /gh-issue-number -->

chris-eibl committed:
minimal changes needed for PGO with clang-cl

chris-eibl committed:
like for MSVC, don't use link time optimization

for _freeze_module in case of clang-cl to speed up the build

chris-eibl committed:
Do not build _freeze_module twice in case of PGO

Speeds up both MSVC and clang-cl builds.

Should most probably done in a separate PR and issue, though.

Fidget-Spinner commented on PCbuild/pythoncore.vcxproj:
```diff
@@ -420,6 +420,7 @@
     <ClCompile Include="..\Modules\blake2module.c">
       <PreprocessorDefinitions Condition="'$(Platform)' == 'x64'">HACL_CAN_COMPILE_SIMD128;%(PreprocessorDefinitions)</PreprocessorDefinitions>
       <PreprocessorDefinitions Condition="'$(Platform)' == 'x64'">HACL_CAN_COMPILE_SIMD256;%(PreprocessorDefinitions)</PreprocessorDefinitions>
+      <AdditionalOptions>/arch:AVX</AdditionalOptions>
```
This should be removed right? It's not always true.

chris-eibl replied:
My clang-cl builds error without that and it does not hurt MSVC AFACT.
It is also used for `Hacl_Hash_Blake2b_Simd256.c` and `Hacl_Hash_Blake2s_Simd128.c`, but there with `Condition="'$(Platform)' == 'x64'`.

I think, I should add that condition here, too.

If in doubt about MSVC, I could also furthermore add `Condition="$(PlatformToolset) == 'ClangCL'`.

chris-eibl commented on PCbuild/pythoncore.vcxproj:
```diff
@@ -716,6 +717,18 @@
     <Delete Files="$(IntDir)pyconfig.h;$(OutDir)pyconfig.h" />
   </Target>
 
+  <Target Name="EnsureClangProfileDir" BeforeTargets="PrepareForBuild"
```
I think, we should not only ensure the profile dir, but empty it.

But most probably only when a full rebuild / clean is done.

chris-eibl committed:
remove /arch:AVX for blake2module.c again

I've previously gotten compile errors from clang, because the needed
intrinsics were not available without that option.

Cannot reproduce anymore. Most probably, because I've upgraded to
Visual Studio 17.13.0 Preview 5.0, which now ships with clang 19.1.1
instead of 18.1.8 and they've done that for compatibility with MSVC?

Anyway, let's keep the PR small :)

chris-eibl committed:
Revert "Do not build _freeze_module twice in case of PGO"

This reverts commit 26fb51f81101fd361411b8e3ae53a3ce0d8bceb1.

Shall be done in a separate PR.

chris-eibl committed:
always clean the profile directory when doing a new clang PGO build

This better matches the behaviour of build.bat in case of MSVC PGO builds.

chris-eibl committed:
Merge remote-tracking branch 'origin/main' into clang-pgo

zooba commented on PCbuild/pythoncore.vcxproj:
```diff
@@ -716,6 +716,26 @@
+     <ItemGroup>
+       <FilesToDelete Include="$(ClangProfileDir)\*.*"/>
+     </ItemGroup>
+    <Delete Files="@(FilesToDelete)" TreatErrorsAsWarnings="true" />
```
Since it's a clean target, let's just `Delete Files="$(ClangProfileDir)..."`. That also will avoid issues in case someone else has already listed files in `FilesToDelete`.

zooba replied:
Probably also fine to move the `Delete` into the above target, so it will always create and delete the contents, doing nothing if there's nothing to be done.

chris-eibl replied:
I tried
```
<Delete Files="$(ClangProfileDir)\*.*" TreatErrorsAsWarnings="true" />
``` 
before I ended up with the `ItempGroup`. Is this what you had in mind?

But since that unfortunately did not delete anything, although it seemed correct to me, I used the  `ItempGroup`.

I also read https://learn.microsoft.com/en-us/visualstudio/msbuild/delete-task?view=vs-2022 during that endeavour, and I also tried `$(ClangProfileDir)\**\*.*` to no avail.

There, that pattern is also warned against - though I don't fear that `ClangProfileDir` will ever be undefined in this case.

Any hints how to proceed?

zooba replied:
Use the item group then, but use a more obscure name and do a `<_WhateverName Remove="@(_WhateverName)" />` first to make sure it's empty.

zooba commented on PCbuild/pyproject.props:
```diff
@@ -70,6 +72,8 @@
       <AdditionalOptions Condition="$(PlatformToolset) == 'ClangCL'">-Wno-deprecated-non-prototype -Wno-unused-label -Wno-pointer-sign -Wno-incompatible-pointer-types-discards-qualifiers -Wno-unused-function %(AdditionalOptions)</AdditionalOptions>
       <AdditionalOptions Condition="$(Configuration) != 'Debug' and $(PlatformToolset) == 'ClangCL'">-flto %(AdditionalOptions)</AdditionalOptions>
+      <AdditionalOptions Condition="$(SupportPGO) and $(Configuration) == 'PGInstrument' and $(PlatformToolset) == 'ClangCL'">-fprofile-instr-generate=$(ClangProfileDir)\default_%m.profraw %(AdditionalOptions)</AdditionalOptions>
+      <AdditionalOptions Condition="$(SupportPGO) and $(Configuration) == 'PGUpdate' and $(PlatformToolset) == 'ClangCL'">-fprofile-instr-use=$(ClangProfileDir)\profdata.profdata %(AdditionalOptions)</AdditionalOptions>
```
I wonder if it may make things cleaner to move these all into their own `<ClCompile Condition="$(PlatformToolset) == 'ClangCL'">` element?

zooba replied:
Or alternatively, move all of the new code into a `pyproject-clangcl.props` and conditionally import that. I think I like that approach best.

chris-eibl replied:
Sounds great, gonna try this :)

That would make the PR slightly greater, since I think you want me to move the few occurences you've added a few years ago as well - correct?

E.g
```
<AdditionalOptions Condition="$(Configuration) != 'Debug' and $(PlatformToolset) == 'ClangCL'">-flto %(AdditionalOptions)</AdditionalOptions>
```

zooba replied:
Yeah, please move everything across. It'll touch more lines, but they should all be simpler, so it'll be easy to review.

zooba commented on PCbuild/pythoncore.vcxproj:
```diff
@@ -716,6 +716,26 @@
     <Delete Files="$(IntDir)pyconfig.h;$(OutDir)pyconfig.h" />
   </Target>
 
+  <Target Name="EnsureClangProfileDir" BeforeTargets="PrepareForBuild"
```
Another trick here would be to add this to pyproject.props, and then do `BeforeTargets="PrepareForClangCLBuild"` here instead of the longer condition:

```
<Target Name="PrepareForClangCLBuild" BeforeTargets="PrepareForBuild" Condition="$(PlatformToolset) == 'ClangCL' />
```

(I'm 99% sure targets "before" a conditional one aren't run if the conditional one is excluded, but worth double checking me on that.)

chris-eibl replied:
I did that, but then it "fired" for every project, but we only want that this is done once, see further down.

zooba replied:
The new `PrepareForClangCLBuild` target should fire for every project, but `EnsureClangProfileDir` would still only be in one place and so it should only run once (though shouldn't it be on the profiling machine? Rather than the build machine? In which case we may need to create it some other way, if the instrumented executables don't do it themselves)

zooba commented on PCbuild/pythoncore.vcxproj:
```diff
@@ -716,6 +716,26 @@
   </Target>
 
+  <Target Name="EnsureClangProfileDir" BeforeTargets="PrepareForBuild"
+          Condition="$(PlatformToolset) == 'ClangCL' and $(Configuration) == 'PGInstrument' and !Exists($(ClangProfileDir))"> 
```
```suggestion
          Condition="$(PlatformToolset) == 'ClangCL' and $(Configuration) == 'PGInstrument'"> 
```

Target conditions are best to be very static, based on input arguments, rather than external state (at least at this level - when using them to process items it's often different).

Making a directory that exists should be a safe no-op.

chris-eibl replied:
Great! Will try. In case it does not work out, I can move the condition into
`<MakeDir Directories="$(ClangProfileDir)" !Exists($(ClangProfileDir)) />`
I had it there in the beginning.

zooba commented on PCbuild/pythoncore.vcxproj:
```diff
@@ -716,6 +716,26 @@
+  <Target Name="MergeClangProfileData" BeforeTargets="PrepareForBuild"
+          Condition="$(PlatformToolset) == 'ClangCL' and $(Configuration) == 'PGUpdate'">
+    <Message Importance="high" Text="Merging clang profile data in $(ClangProfileDir)" />
+    <Exec
```
Bearing in mind that we may collect the profile on a different machine from where we run the PGUpdate stage, do we need to merge on the profiling machine? Or on the build machine?

This merges on the build machine, same as for MSVC. We'd probably need it to be its own top-level target that can be run directly if it should be on the profiling machine.

chris-eibl replied:
How would I make it a top-level target?

Is there anything to do now, or is this for future consideration?
I think I did not 100% understand this.

IIUC, the same would apply to the existing MSVC pgo build

There currently building and profiling can only be done at the same machine AFAIU.

zooba replied:
Basically, a top-level target just isn't referenced by any others, and doesn't specify BeforeTargets or AfterTargets. So the only way to run it is `msbuild /t:NameOfTarget`.

For MSVC we already profile on a separate machine for ARM64, so it's definitely possible to build in one place and profile in another. Collecting the profile doesn't rely on anything in MSBuild though - it all gets embedded during the PGInstrument phase. I'm not sure whether the same can be said of this command or not.

It's a bit complex since we use it a couple of different ways, but you can see the steps for the split PGO build at https://github.com/python/release-tools/blob/master/windows-release/build-steps-pgo.yml

chris-eibl replied:
Ah, I see.  So basically, these are the steps, as far as I understand it:
- `build machine`: build.bat -c PGInstrument
   The instrumented binaries are created and published (`$(Build.BinariesDirectory)\bin\$(Arch)\instrumented` ). In the MSVC case, we have `*.pgd` files here in the instrumented folder. This is even easier in the clang case, since no such files are generated/needed.
   
- `profile machine`:
`$(Build.BinariesDirectory)\bin\$(Arch)\instrumented` is fetched and then
```
$exe = (gci "$(Build.BinariesDirectory)\bin\$(Arch)\instrumented\$(PythonExePattern)" | select -First 1)
& $exe -m test --pgo
``` 
is executed.  This the "Collecting profile data" step. Profile data is written. In case of MSVC, the `*!N.pgc` files are written just "next to the binaries". This whole folder is then re-published again. Therefore, in the MSVC case we do not have to care about the profile dir at all. It is implicitely the "instrumented" folder.

- back on `build machine`:
  fetches `$(Build.BinariesDirectory)\bin\$(Arch)\instrumented`
  build.bat -c PGUpdate
  `<Target Name="CopyPGCFiles">` takes care to copy the `!*.pgc` and `*.pgd` files. This is done per `TargetName`:
  
  `<_PGCFiles Include="$(OutDir)instrumented\$(TargetName)!*.pgc" />`

  So e.g. when building `_asyncio.pyd`,  only `_asyncio!*.pgc` and `_asyncio.pgc` are copied here to `$(OutDir)`.
  Here, no "concurrent writes" to the same output can happen.
  
I think, it would be best to find out how to tell the instrumented clang binaries, that they should generate their respective `*.profraw` file "next to themselves". 

Maybe https://clang.llvm.org/docs/UsersManual.html#profiling-with-instrumentation can help here. There are many options, but I think none easily does the task.

Then everything could be done pretty much the same as in the MSVC case. Except that all those `*.profraw` files must be merged into the one-and-only `final.profdata` in one go, before the PGUpdate step starts.

Maybe our Linux PGO builds have solved the build / profile on different machines?

At least the makefiles seem damn simple. Just (ab)use the current working dir:
```
LLVM_PROF_MERGER= /usr/bin/../lib/llvm-18/bin/llvm-profdata merge -output="$(shell pwd)/code.profclangd" "$(shell pwd)"/*.profclangr 
```

zooba commented on PCbuild/pythoncore.vcxproj:
```diff
@@ -716,6 +716,26 @@
     <Delete Files="$(IntDir)pyconfig.h;$(OutDir)pyconfig.h" />
   </Target>
 
+  <Target Name="EnsureClangProfileDir" BeforeTargets="PrepareForBuild"
```
I suspect we need these steps in every project, and not just python3x.dll? At least those that support PGO (which is an existing property in most projects).

If they need to run exactly once for the entire build, it probably needs to be a new .proj file entirely that's scheduled last (or run manually). Otherwise, they should go in `pyproject.props` or `pyproject-clangcl.props` as suggested above.

chris-eibl replied:
This is a sketch of what I had to achive:

First, the PGInstrument step is done. Nothing is needed there, yet.

We need the profile dir before we start the `pgo_job` in build.bat
```
call "%dir%\..\python.bat" %pgo_job%
``` 
This will generate all the `default_*.profraw` files.
Then they must be merged into the final `profdata.profdata` file.
This has to be done before the PGUpdate build starts.
In the PGUpdate build itself, this file is then read by all projects.

Thus, for me it was the easiest way to put it into `pythoncore.vcxproj`, because this is run first after the freeze module without any other projects being built in parallel. But I admit, it's quite obfuscated there.

If I understand you correctly, I could create a new project (*.vcxproj) like the freeze_project, that I then call in the `pcbuild.proj` right after the `"@(FreezeProjects)"` that does in case of the
- PGInstrument step: ensure that there is an empty `$(ClangProfileDir)`
- PGUpdate step:  the merge

Actually, that's the only task to solve here, which I did so complicated, because I did not want to touch `build.bat` and put "too much knowledge in there" like `ClangProfileDir` and `LLVMInstallDir`
```
ensure-empty-ClangProfileDir
call "%dir%\..\python.bat" %pgo_job%
$(LLVMInstallDir)\bin\llvm-profdata.exe" merge -output=$(ClangProfileDir)\profdata.profdata $(ClangProfileDir)\default_*.profraw'
```

zooba replied:
If the resulting filename is known, then you can set up a target with `Inputs="default_*.profraw" Outputs="profdata.profdata"` and let it run in every project. That way the files are guaranteed to be merged if they're newer than the output, but it won't do it if it isn't necessary.

chris-eibl replied:
I tried that, too. But must have done something wrong. Because it first merged for pythoncore.vcxproj, where luckily no other projects build in parallel. So the data was once merged and should have been "up do date".

Then the other *.vcxproj where processed, many in parallel. Each of them still wanted to merge, and this does not work in parallel, since many invocations of `llvm-profdata.exe` cannot concurrently write into the same file.

Must have messed up something here, because `Outputs="profdata.profdata"` was already newer than the inputs :(

Will try again, and if I make the same mistake again, maybe you'll spot it :)

zooba replied:
I think I normally end up with item groups. There's probably another example somewhere in the projects where I got it working - have a look for other `Outputs=` specifiers.

zooba commented on PCbuild/_freeze_module.vcxproj:
```diff
@@ -92,6 +92,7 @@
       <AdditionalIncludeDirectories>$(IntDir);%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
       <Optimization>Disabled</Optimization>
       <WholeProgramOptimization>false</WholeProgramOptimization>
+      <AdditionalOptions Condition="$(PlatformToolset) == 'ClangCL'">%(AdditionalOptions) -fno-lto</AdditionalOptions>
```
I wonder if there's a way we can put this into a shared props file and test the `WholeProgramOptimization` property, in case it gets added elsewhere? There probably isn't, I think the import ordering will be wrong... though maybe in a PrepareForBuild target?

Just thinking out loud. No change needed here unless you have a brilliant idea.

chris-eibl replied:
Boah, sounds great, but far above my knowledge about msbuild/vcxproj etc.

You've given me already so much valuable input, going to digest that first :)

zooba reviewed (commented):
More proposed changes than I expected, but it's mostly organisation :) I'm impressed it's actually this simple.

chris-eibl committed:
Merge branch 'main' into clang-pgo

chris-eibl committed:
blurb it

chris-eibl committed:
Adding myself to ACKS

chris-eibl committed:
extract pyproject-clangcl.props

chris-eibl committed:
Merge branch 'python:main' into clang-pgo

chris-eibl committed:
move MergeClangProfileData to pyproject-clangcl.props

and make it a target with inputs and outputs

chris-eibl committed:
rename RequirePGCFiles to RequireProfileData

because the name is too MSVC specific

chris-eibl commented on PCbuild/pyproject-clangcl.props:
```diff
@@ -0,0 +1,39 @@
+    <ClCompile>
+      <AdditionalOptions>-Wno-deprecated-non-prototype -Wno-unused-label -Wno-pointer-sign -Wno-incompatible-pointer-types-discards-qualifiers -Wno-unused-function %(AdditionalOptions)</AdditionalOptions>
+      <AdditionalOptions Condition="$(Configuration) != 'Debug'">-flto %(AdditionalOptions)</AdditionalOptions>
+      <AdditionalOptions Condition="$(SupportPGO) and $(Configuration) == 'PGInstrument'">-fprofile-instr-generate=$(OutDir)$(TargetName)_%m.profraw %(AdditionalOptions)</AdditionalOptions>
```
`$(OutDir)$(TargetName)_%m.profraw` is still an absolute path on the build machine, so this won't work in case the PGO task is run remotely.

IMHO, like in the MSVC case, the build host shall expect that the profile data comes next to the instrumented binaries. That's the 'current contract'. Let's not break it :) Then, in the PGUpdate phase, everything else stays the same and in sync with the MSVC way.

I am thinking of some options how to achieve this:
- finding a way how to let clang automagically place the profile data next to the binary, like done by MSVC. That would be soooo convenient. I did read https://clang.llvm.org/docs/UsersManual.html#profiling-with-instrumentation again, nothing seems to fit.
-  using just `-fprofile-instr-generate` without a parameter and using the environment variable `LLVM_PROFILE_FILE` on the PGO host seems tempting. We'd lose the `$(TargetName)` part, though :( And it needs some default handling for the non-remote case anyway, so therefore
- I propose to introduce something like `CLANG_PROFILE_PATH` and default to `$(OutDir)`. We'd then pass in `/p:CLANG_PROFILE_PATH=<path-to-binaries-on-remote-host>` and we use `$(CLANG_PROFILE_PATH)$(TargetName)_%m.profraw`.

zooba replied:
We can copy the profiled data back into the same location (relative to `$(OutDir)`) for the PGUpdate step, that's not a problem. We just can't assume that we can run the profiling job itself from the build scripts.

But it doesn't actually look like we need a full path for `-fprofile-instr-generate`? So setting it to `-fprofile-instr-generate=$(TargetName)_%m.profraw` should be okay, right?  That part just gets compiled into the binary, and presumably is going to be interpreted as relative to either the current directory or the executable, both of which we can control when running a job. The path in the below item for PGUpdate can be a full path, because we know we've created that file locally.

chris-eibl replied:
>  full path for -fprofile-instr-generate

Correct, it can be a relative path and then is relative to the current working directory.

> We can copy the profiled data back

If my analysis https://github.com/python/cpython/pull/129907#discussion_r1955293669 of [build-steps-pgo.yml](https://github.com/python/release-tools/blob/master/windows-release/build-steps-pgo.yml) is correct:

That's what I'd like to achieve with the `$(CLANG_PROFILE_PATH)$(TargetName)_%m.profraw` approach.
If not passing it in the "regular" `build.bat --pgo` case, `$(CLANG_PROFILE_PATH)` will default to `$(OutDir)`  and everything is working just like now.

Using 
```
 build.bat -c PGInstrument /p:CLANG_PROFILE_PATH=<relative-path-to-instrumented-dir-on-remote-host>
``` 
on the build host, will assure that the `$(TargetName)_%m.profraw` files are then generated into the `$(Build.BinariesDirectory)\bin\$(Arch)\instrumented` on the PGO host, without any other modifications in the yml
```
      $exe = (gci "$(Build.BinariesDirectory)\bin\$(Arch)\instrumented\$(PythonExePattern)" | select -First 1)
      & $exe -m test --pgo
```
They are then already out-of-the-box published.

Back on the build host, the 
```
build.bat -c PGUpdate
````
will continue to work (unmodified), too.

I thought of that `CLANG_PROFILE_PATH`, because
- just passing it in the PGInstrument step of the remote PGO scenario is the only necessary change to make it work
- all the rest of [build-steps-pgo.yml](https://github.com/python/release-tools/blob/master/windows-release/build-steps-pgo.yml) remains unchainged
- "simple" change
  - does not clutter the current working directory in the local `build.bat --pgo` case.
  - no further copying/cleaning/etc. needed
  - different local clang-cl PGO builds (32bit / 64bit) don't interfere, since the artifacts / object files are already separated in a clean way and thus I opted for  `$(OutDir)instrumented\$(TargetName)_*.profraw` (like in the MSVC case).

I tried this in https://github.com/python/cpython/pull/129907/commits/9db1a297d95574cc3114854c8c427736d9521917 and for me
- the regular clang-cl PGO build
```
build.bat --pgo "/p:PlatformToolset=ClangCL" "/p:PreferredToolArchitecture=x64"` 
```
- and a "faked" remote PGO build using
```
build.bat --pgo "/p:PlatformToolset=ClangCL" "/p:PreferredToolArchitecture=x64" "/p:CLANG_PROFILE_PATH=amd64\instrumented"
```
to simulate a relative path on the remote PGO host both worked.

zooba replied:
Okay, we'd better write up some docs on how to use this option properly. A new section in PCbuild/readme.txt is probably appropriate (covering all of building with clang, but I think it can all be pretty brief, provided it includes the explanation for this part).

zooba commented on PCbuild/pyproject-clangcl.props:
```diff
@@ -0,0 +1,39 @@
+  <Target Name="MergeClangProfileData" BeforeTargets="PrepareForBuild"
+          Condition="'$(SupportPGO)' and $(Configuration) == 'PGUpdate'"
+          Inputs="@(_profrawFiles)"
+          Outputs="@(_profrawFiles->'$(OutDir)instrumented\profdata.profdata')">
```
```suggestion
          Outputs="$(OutDir)instrumented\profdata.profdata">
```

I'm 99% sure this should work (different numbers of inputs vs. outputs is fine), but if you tried it and it didn't then no worries.

chris-eibl replied:
Haven't tried before but can confirm your suggestion works!

zooba commented on PCbuild/pyproject-clangcl.props:
```diff
@@ -0,0 +1,39 @@
+          Inputs="@(_profrawFiles)"
+          Outputs="@(_profrawFiles->'$(OutDir)instrumented\profdata.profdata')">
+    <Exec
+      Command='"$(LLVMInstallDir)\bin\llvm-profdata.exe" merge -output=$(OutDir)instrumented\profdata.profdata $(OutDir)instrumented\*_*.profraw' />
```
```suggestion
      Command='"$(LLVMInstallDir)\bin\llvm-profdata.exe" merge -output="$(OutDir)instrumented\profdata.profdata" "$(OutDir)instrumented\*_*.profraw"' />
```

zooba commented on PCbuild/pyproject-clangcl.props:
```diff
@@ -0,0 +1,39 @@
+  </Target>
+
+  <Target Name="CleanClangProfileData" BeforeTargets="Clean">
+    <Delete Files="@(_profrawFiles->'$(OutDir)instrumented\%(Filename)%(Extension)')" TreatErrorsAsWarnings="true" />
```
```suggestion
    <Delete Files="@(_profrawFiles)" TreatErrorsAsWarnings="true" />
```

Unless the item group gets updated somewhere I didn't notice?

chris-eibl replied:
Ah - this works as well! Stumbled over that, too, and already simplified to `@(_profrawFiles->'%(FullPath)')`, but cannot beat your suggestion :)

chris-eibl committed:
Apply suggestions from code review

Co-authored-by: Steve Dower <steve.dower@microsoft.com>

chris-eibl committed:
Use -Wno-profile-instr-unprofiled in the PGUpdate case

chris-eibl commented on PCbuild/pyproject-clangcl.props:
```diff
@@ -32,7 +32,7 @@
       <AdditionalOptions Condition="$(Configuration) != 'Debug'">-flto %(AdditionalOptions)</AdditionalOptions>
       <AdditionalOptions Condition="$(SupportPGO) and $(Configuration) == 'PGInstrument'">-fprofile-instr-generate=$(OutDir)$(TargetName)_%m.profraw %(AdditionalOptions)</AdditionalOptions>
-      <AdditionalOptions Condition="$(SupportPGO) and $(Configuration) == 'PGUpdate'">-fprofile-instr-use=$(OutDir)instrumented\profdata.profdata %(AdditionalOptions)</AdditionalOptions>
+      <AdditionalOptions Condition="$(SupportPGO) and $(Configuration) == 'PGUpdate'">-fprofile-instr-use=$(OutDir)instrumented\profdata.profdata -Wno-profile-instr-unprofiled %(AdditionalOptions)</AdditionalOptions>
```
Use `-Wno-profile-instr-unprofiled` to silence warnings like
```
TRACKEDEXEC : warning : no profile data available for file "filter_buffer_encoder.c" [-Wprofile-instr-unprofiled]
``` 

I think we are not interested in 33 additional warnings like this?
clang-cl already issues a bunch of warnings, so let's not increase them that much further.

chris-eibl replied:
So this seems ok and since it is not set to resolved, yet, I assume I can (shall?) do it?
Sorry, the dev guide is not clear on that.
My feeling says: if I am quite confident, that it is resolved, I can klick the button, if in doubt, rather keep it open?

chris-eibl committed:
introduce CLANG_PROFILE_PATH

chris-eibl committed:
update readme.txt

Fidget-Spinner commented on PCbuild/readme.txt:
```diff
@@ -52,6 +52,42 @@ Release
+--------------------------------
+
+See https://learn.microsoft.com/en-us/cpp/build/clang-support-msbuild?view=msvc-170
+how to install and use clang-cl bundled with Microsoft Visual Studio.
```
```suggestion
See https://learn.microsoft.com/en-us/cpp/build/clang-support-msbuild?view=msvc-170
for how to install and use clang-cl bundled with Microsoft Visual Studio.
```

Fidget-Spinner commented on PCbuild/readme.txt:
```diff
@@ -52,6 +52,42 @@ Release
+All other build.bat options remain to work as with MSVC, so this
+will create a 64bit release binary. PreferredToolArchitecture is needed,
+because msbuild by default selects the 32bit architecture of the toolset,
+which uses -m32 as the default target architecture.
```
```suggestion
All other build.bat options continue to work as with MSVC, so this
will create a 64bit release binary. PreferredToolArchitecture is needed,
because msbuild by default selects the 32bit architecture of the toolset,
which uses -m32 as the default target architecture.
```

Fidget-Spinner commented on PCbuild/readme.txt:
```diff
@@ -52,6 +52,42 @@ Release
+and use 18.1.8 in the above example, too.
+
+Even --pgo works out of the box, except you do want to run the PGO task
+on a different host than the build host. In this case you must pass
```
```suggestion
Even --pgo works out of the box. However, if you want to run the PGO task
on a different host than the build host, you must pass
```

Fidget-Spinner reviewed (commented):
Just some small comments for the instructions.

chris-eibl committed:
Apply suggestions from code review

Co-authored-by: Ken Jin <kenjin4096@gmail.com>

zooba commented on PCbuild/readme.txt:
```diff
@@ -52,6 +52,42 @@ Release
+Building Python using Clang/LLVM
+--------------------------------
+
+See https://learn.microsoft.com/en-us/cpp/build/clang-support-msbuild?view=msvc-170
```
```suggestion
See https://learn.microsoft.com/cpp/build/clang-support-msbuild
```

zooba commented on PCbuild/readme.txt:
```diff
@@ -52,6 +52,42 @@ Release
+All other build.bat options continue to work as with MSVC, so this
+will create a 64bit release binary. PreferredToolArchitecture is needed,
+because msbuild by default selects the 32bit architecture of the toolset,
+which uses -m32 as the default target architecture.
```
If this is genuinely required, let's add a `<PreferredToolArchitecture Condition="$(PROCESSOR_ARCHITECTURE) == 'AMD64'">x64</PreferredToolArchitecture>` to the clangcl.props. I prefer to not override MSVC defaults for this property, but if Clang can't handle cross-compiling, then provided we know that it's been requested we may as well override defaults.

chris-eibl replied:
TBH, I don't know why I need it when building with msbuild. I do not need it in case of the IDE or when explicitely using a custom installed toolset (most probably because I've always installed 64bit versions of those). I like your suggestion and will try it out - looks promising!

chris-eibl replied:
> looks promising!

But did not work out: most probably, we're "too late" when setting that, meaning that these lines in `Microsoft.Cpp.ClangCl.Common.props` are already "resolved":

```
    <_DefaultLLVMInstallDir Condition="'$(_DefaultLLVMInstallDir)' == '' AND '$(PreferredToolArchitecture)' == 'arm64'">$(VsInstallRoot)\VC\Tools\Llvm\ARM64</_DefaultLLVMInstallDir>
    <_DefaultLLVMInstallDir Condition="'$(_DefaultLLVMInstallDir)' == '' AND '$(PreferredToolArchitecture)' == 'x64'">$(VsInstallRoot)\VC\Tools\Llvm\x64</_DefaultLLVMInstallDir>
    <_DefaultLLVMInstallDir Condition="'$(_DefaultLLVMInstallDir)' == '' AND '$(PreferredToolArchitecture)' != 'x64'">$(VsInstallRoot)\VC\Tools\Llvm</_DefaultLLVMInstallDir>
    <LLVMInstallDir Condition="'$(LLVMInstallDir)' == ''">$(_DefaultLLVMInstallDir)</LLVMInstallDir>
```

Anyway, using the `PreferredToolArchitecture` as a way to steer the architecture of the generated binaries was a bad idea. So to get a 32bit binary, one would have to use `"/p:PreferredToolArchitecture=x86"` :-O

`clang-cl` of course can cross-compile, so let's explicitely set the architecture based on $(Platform).

zooba commented on PCbuild/readme.txt:
```diff
@@ -52,6 +52,42 @@ Release
+Setting the major version is enough, although you can be specific
+and use 18.1.8 in the above example, too.
+
+Even --pgo works out of the box. However, if you want to run the PGO task
```
```suggestion
Use the --pgo option to build with instrumented profiling. However, if you want to run the PGO task
```

Let's not sound too surprised that it works 😉 Also would be good to replace "instrumented profiling" with whatever word will help someone who knows Clang already understand exactly what we're using.

chris-eibl replied:
They use PGO as well, quote from  https://llvm.org/docs/HowToBuildWithPGO.html
```
PGO (Profile-Guided Optimization)
``` 
Shall I link there or to https://clang.llvm.org/docs/UsersManual.html#profiling-with-instrumentation?
There, they frequently use the term `PGO`, too.

zooba replied:
If the term is the same, that's fine. Main thing is to use this sentence to bridge any terminology gaps, so if they don't exist, then let's just be direct.

zooba commented on PCbuild/readme.txt:
```diff
@@ -52,6 +52,42 @@ Release
+
+Even --pgo works out of the box. However, if you want to run the PGO task
+on a different host than the build host, you must pass
+"/p:CLANG_PROFILE_PATH=<relative-path-to-instrumented-dir-on-remote-host>"
+in the PGInstrument step to make sure the profile data is generated
+into the instrumented directory when running the PGO task.
```
I think this needs a worked example. Just from this description, I couldn't confidently specify the parameter and trust that I got it right.

zooba reviewed (commented):
Couple of readme suggestions, but I think we're nearly there!

chris-eibl committed:
Apply suggestions from code review

Co-authored-by: Steve Dower <steve.dower@microsoft.com>

chris-eibl committed:
credit zooba

chris-eibl committed:
address review comments regarding PGO

chris-eibl committed:
Explicitely set the architecture based  on $(Platform)

to be independent from the selected architecture
or the PlatformToolset

zooba commented on PCbuild/pyproject-clangcl.props:
```diff
@@ -39,6 +39,8 @@
     <ClCompile>
       <AdditionalOptions>-Wno-deprecated-non-prototype -Wno-unused-label -Wno-pointer-sign -Wno-incompatible-pointer-types-discards-qualifiers -Wno-unused-function %(AdditionalOptions)</AdditionalOptions>
+      <AdditionalOptions Condition="'$(Platform)' == 'Win32'">-m32 %(AdditionalOptions)</AdditionalOptions>
+      <AdditionalOptions Condition="'$(Platform)' == 'x64'">-m64 %(AdditionalOptions)</AdditionalOptions>
```
Is there an ARM64 option as well?

zooba replied:
I'm glad you found this option, though! Much neater way to do it (and a bit embarrassing that the built-in targets don't do it...).

chris-eibl replied:
> Is there an ARM64 option as well?

I think so, yes. I see some hits in the net, and did a `clang-cl /?`, which e.g. reveals option `/arm64EC`.

But I do not know what to actually write here in case of `'$(Platform)'=='ARM64'` - and more importantly how to test it, since I have no arm based device.

I think we should leave that for another PR of someone who does?

zooba replied:
Sure. @brandtbucher might be that someone. But at the conditions here will just not set an option on that platform and so it should work the same as today.

zooba commented on PCbuild/readme.txt:
```diff
@@ -61,29 +61,32 @@ You can use the IDE to switch to clang-cl for local development,
+E.g., if you place the instrumented binaries into the folder
+"workdir/instrumented" and then run the PGO task using "workdir"
+as the current working directory, the usage is
+"/p:CLANG_PROFILE_PATH=workdir/instrumented"
```
```suggestion
"/p:CLANG_PROFILE_PATH=instrumented"
```

It goes into `workdir` already, right? Or do we actually need to repeat it?

chris-eibl replied:
My bad, sure, that was brain damage on my side - I should go to bed now :)

chris-eibl committed:
Update PCbuild/readme.txt

Co-authored-by: Steve Dower <steve.dower@microsoft.com>
